### PR TITLE
Greatest Common Divisor (Euclidian Algorithm) in Dart

### LINKED ISSUE #6300: Greatest Common Divisor (Euclidian Algorithm) in Dart
Greatest Common Divisor in Dart

### PR DESCRIPTION
**Pull request by ashuvssut**
Greatest Common Divisor (Euclidian Algorithm) in Dart

I have written the comments to give a explanation and a proof for the Euclidian algorithm that states that `the remainder a' of a/b is also divisible by the GCD of a and b (i.e. d)` This single fact is used in the algorithm. The Statement might be confusing but my proof for that will make it easy to understand

My examples explanations illustrates how Time Complexity is `O(logn)` 😎

Hope you ll like it 😁

review link https://replit.com/@AshutoshKhandua/SaneOrderlyForm

closes #6300

### CODE REVIEW

**Review by atarax665 — APPROVED** [review]

File: Dart/maths/Euclidean_Algorithm_to_find_GCD.dart [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    Explanation
+    gcd(100,70)
+    =gcd(70,30)
+    =gcd(30,10)
+    =gcd(10,0)
+
+    here we get the gcd=10.
+    Observation:- all the remainders and those numbers a & b have the same GCD. 
+    i.e., 100, 70, 30, 10, 0 have gcd=10
+
```
Comment (ankitaggarwal23) [inline Dart/maths/Euclidean_Algorithm_to_find_GCD.dart:106]: Remove extra test case

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]

File: Dart/maths/Euclidean_Algorithm_to_find_GCD.dart [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  stdout.write("Enter the first number: ");
+  int a = int.parse(stdin.readLineSync());
+  stdout.write("Enter the second number: ");
+  int b = int.parse(stdin.readLineSync());
+
+  /*
+    here a>b or b>a does not matter
+    because a will be the greater in the algo's 1st recursion. for example:-
+    gcd(25,100)
+    = gcd(100,25) here 25%100=25
```
Comment (ashuvssut) [inline Dart/maths/Euclidean_Algorithm_to_find_GCD.dart:62]: Example 1 is referred here. this example 1 is important

it shows why gcd(a,b) ultimately becomes gcd(b,a) if a<b

File: Dart/maths/Euclidean_Algorithm_to_find_GCD.dart [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+    a/d and b/d are in P/Q form where Q=1 since d is their GCD
+    
+    =>  a/d   -   q*(b/d) =   a'/d
+    =>        X           =   a'/d where X is a positive integer
+
+    this also implies that a'/d is also an integer (in p/q form where q=1)
+
+    so it is proved that the remainder a' of  a/b is also divisible by the gcd of a and b (i.e. d)
+    i.e. a,b,d have common gcd (You are going to get it at the end of calculation ---- consider example2)
```
Comment (ashuvssut) [inline Dart/maths/Euclidean_Algorithm_to_find_GCD.dart:25]: Example 2 is reffered here. 

It shows how a,b,d have a common gcd d
gcd(100,70)
    =gcd(70,30)
    =gcd(30,10)
    =gcd(10,0)

it shows how problem is solved in O(logn)

File: Dart/maths/Euclidean_Algorithm_to_find_GCD.dart [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+*/
+
+import 'dart:io';
+
+int gcd(int a, int b) {
+  if (b == 0) return a;
+
+  return gcd(b, a % b);
+  /* 
+    Why return gcd(b, a % b) and not gcd(a % b, b)???
```
Comment (ashuvssut) [inline Dart/maths/Euclidean_Algorithm_to_find_GCD.dart:36]: Example 2 is reffered here.

it supports why in subsequent stack-frame we return gcd(b, a % b) and not gcd(a % b, b)

gcd(100,70)
=gcd(70,30)
=gcd(30,10)
=gcd(10,0)

**Comment by ashuvssut** [comment]
@ankitaggarwal23 should i still delete the test case?

Both are important and serve as an example for my explanation i gave for the algorithm (see my reviews on those two lines}

Please let me know. I will do what you want here

**Review by ankitaggarwal23 — APPROVED** [review]

**Pull request merged**